cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,271 @@
+
+    :param path: string representing a Windows path. The root drive must exist,
+    although the path is not required to.
+    """
+    out, _ = subp.subp([WSLPATH_CMD, "-au", path])
+    return PurePath(out.rstrip())
+
+
+def cmd_executable() -> PurePath:
+    """
+    Returns the Linux path to the Windows host's cmd.exe.
+    """
+
+    mounts = mounted_win_drives()
+    if not mounts:
+        raise IOError("Windows drives are not mounted.")
+
+    # cmd.exe path is being stable for decades.
+    candidate = "%s/Windows/System32/cmd.exe"
+    for mnt in mounts:
+        cmd = candidate % mnt
+        if not os.access(cmd, os.X_OK):
+            continue
+
+        LOG.debug("Found cmd.exe at <%s>", cmd)
+        return PurePath(cmd)
+
+    raise IOError(
+        "Couldn't find cmd.exe in any mount point: %s" % ", ".join(mounts)
+    )
+
+
+def cloud_init_data_dir() -> PurePath:
+    """
+    Returns the Windows user profile directory translated as a Linux path
+    accessible inside the current WSL instance.
+    """
+    cmd = cmd_executable()
+
+    # cloud-init runs too early to rely on binfmt to execute Windows binaries.
+    # But we know that `/init` is the interpreter, so we can run it directly.
+    # See /proc/sys/fs/binfmt_misc/WSLInterop[-late]
+    # inside any WSL instance for more details.
+    home, _ = subp.subp(["/init", cmd.as_posix(), "/C", "echo %USERPROFILE%"])
+    home = home.rstrip()
+    if not home:
+        raise subp.ProcessExecutionError(
+            "No output from cmd.exe to show the user profile dir."
+        )
+
+    win_profile_dir = win_path_2_wsl(home)
+    seed_dir = os.path.join(win_profile_dir, ".cloud-init")
+    if not os.path.isdir(seed_dir):
+        raise FileNotFoundError("%s directory doesn't exist." % seed_dir)
+
+    return PurePath(seed_dir)
+
+
+def candidate_user_data_file_names(instance_name) -> List[str]:
+    """
+    Return a list of candidate file names that may contain user-data
+    in some supported format, ordered by precedence.
+    """
+    distribution_id, version_id, _ = util.get_linux_distro()
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, version_id),
+        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "default.user-data",
+    ]
+
+
+DEFAULT_INSTANCE_ID = "iid-datasource-wsl"
+
+
+def load_instance_metadata(cloudinitdir: PurePath, instance_name: str) -> dict:
+    """
+    Returns the relevant metadata loaded from cloudinit dir based on the
+    instance name
+    """
+    metadata = {"instance-id": DEFAULT_INSTANCE_ID}
+    metadata_path = os.path.join(
+        cloudinitdir.as_posix(), "%s.meta-data" % instance_name
+    )
+    try:
+        metadata = json.loads(util.load_binary_file(metadata_path))
```

Comment:
In further discussion with the team, we came to the conclusion that we don't want to force JSON on this file format, but instead rely on YAML format like DataSourceNoCloud.
YAML also has the benefit  of not requiring as many characters to express the simple dictionary values that is probably less prone to errors as well:
`instance-id: some-uuid`

@CarlosNihelton sorry for thrashing on this. 
Can we instead use `util.load_yaml?


Also let's get some unittest coverage on load_instance_metadata. We'll need to update docs for the meta-data file to simplify the value to `instance-id: some-uuid` from the JSON equivalent.

I'll push this suggested diff for review on this branch to short-circuit the work on your end if you agree.